### PR TITLE
Adding Python environments with README instructions; starting to extract JIRA data from a xml file

### PR DESCRIPTION
**Pull request by IanSinColorado**
Check out the instructions in the README and see if the code segments work for reproducing the environment(s)

### CODE REVIEW

**Comment by MattF-NSIDC** [comment]
Hey, @IanSinColorado ! Hope you don't mind my comments. The new repo appeared in my notifications and I decided to take a peek :)

In my experience the easiest way to deal with Conda environments is hand-maintaining an `environment.yml` which contains only your top-level dependencies (including both dev and runtime dependencies) and no sub-dependencies (i.e. include `beautifulsoup4`, but not `urllib3`).

`conda env export` is producing a record of your exact environment at that point in time, which may not be installable on other platforms and may even stop working next month on the same computer. If you instead hand-maintain an `environment.yml` which uses [compatible release](https://peps.python.org/pep-0440/#compatible-release) specification (e.g. `beautifulsoup4 ~=4.12`) and contains only things you directly `import`, then environment maintenance will be much easier over time. If you take this approach, I recommend avoiding `conda install <some-package>`, because that won't update your `environment.yml`. Instead, you could add `some-package` to your `environment.yml`, then do `conda env update` to apply the change to your environment.

An environment file with exact pins like you have is really useful if you want to build a docker image or run CI, but if you're not doing those things you may not need one at all. On NSIDC projects, we typically call those "lock files" following the nomenclature from NPM's `package-lock.json`, and often save them as `environment-lock.yml`. Some things to be aware of when using `conda env export`:

* the YAML outputs are platform-specific, and if you're using a Mac, that lock file may not work in GitHub Actions' Linux runners or on other developers' computers. For that, there's a tool called `conda-lock` which can create multi-platform lock files.
* the YAML outputs are specific to your personal computer; the last line of the YAML is `prefix: <some-path-on-your-computer>` and can (should?) be omitted. A one-liner: `conda env export | grep -v "^prefix:" > environment-lock.yml`
* the YAML outputs will contain your channel configuration from your `~/.condarc` file, which may not match the project's needs

[Here's](https://github.com/nsidc/sipn-reanalysis-plots/blob/main/environment.yml) an example of an `environment.yml` that's maintained with the above practices. The same repo also has an [`environment-lock.yml` file](https://github.com/nsidc/sipn-reanalysis-plots/blob/main/environment-lock.yml) which we produce using `conda env export` on a Linux machine so it can be built in GitHub Actions.

I'd also recommend against including your individual development dependencies (e.g. `vim`, `vscode`) in a project conda environment, and instead install them with Homebrew/Apt/whatever, as these may conflict with installed tools on other people's computers! Shared dev dependencies like `pytest` or `pre-commit` are more appropriate for `environment.yml`.

**Comment by IanSinColorado** [comment]
I just added a commit with a manual environment.yml file. I tested it by starting to parse through the test JIRA issues XML file

**Comment by MattF-NSIDC** [comment]
Looks great! Good call on writing up a quick test :sunglasses: I have a couple small suggestions:

* The `anaconda` channel is OK, but it's not worth using compared to the `conda-forge` channel. Packages on `anaconda` are built in private by ContinuumIO with little community input or control, while `conda-forge` packages are community maintained and built in the open from "feedstocks", which are GitHub repos containing packaging "recipes" only. I love this because when I need a new version that's not available, ContinuumIO won't take my call about updating the `anaconda` channel, but I can easily [open a feedstock PR](https://github.com/conda-forge/quarto-feedstock/pull/6) to get the version I need released onto `conda-forge`. I suggest replacing `anaconda` with `conda-forge` and adding the channel `nodefaults` to the end of the list. This prevents conda from using the `anaconda` channel implicitly, which can be annoying. I'd even suggest creating or updating a `~/.condarc` that sets the channels for your user to always be `conda-forge` and `nodefaults`.
* [beautifulsoup4](https://anaconda.org/conda-forge/beautifulsoup4) and [lxml](https://anaconda.org/conda-forge/lxml) are both available on `conda-forge`, so those can be pulled out of the `pip` section and into the main dependency list; you don't need the `pip` section at all :)

**Comment by IanSinColorado** [comment]
I have completed the issue of creating a reproducible conda environment. There are instructions in the README if you want to test it out. I think we need to merge this pull request and start a new branch/pull request, but in the meantime, I have started to load JIRA issue data from the jiraIssuesCryo.xml file. 
So far, I have functions to extract the title, JIRA issue link, and description (raw and pretty). I still need to extract the labels, status, and assignees from the XML file. Finally, I need to check if any other JIRA issue fields can be migrated to GitHub.

File: README.md [hunk 1]
```diff
@@ -24,6 +24,23 @@ nsidc@nsidc.org for more information.
 
 * JIRA Ticket XML
 * Access to GitHub repository
+* Anaconda and Conda installed locally
+
+## Conda Environment Setup
+This section will serve as a guide to set up the proper environment with the required libraries.
+
+1. Clone this repository into a directory of your choosing
+2. Open the Anaconda Prompt (do a search on your computer for 'Anaconda Prompt' or open any terminal that can use the `conda` keyword; check using `conda --version` to see if a version pops up)
```
Comment (jwfuller) [inline README.md]:
```suggestion
2. Verify to have [conda installed](https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html): open a terminal and run `conda --version`. If a version number is returned, you have conda or miniconda installed.
```

File: GitHubAPI.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+issues = JL.JIRALoader('../testJIRA.xml')
+
+# Getting the issues loaded into the JIRALoader object
+issues.get_issue_data()
+
+REPO_OWNER = 'iansincolorado'
+REPO_NAME = 'pythonapis'
+
+# Only way of authenticating at the current moment is with GitHub PAT tokens, which are deleted from your GitHub account if it is found in a public repository
+PAT = ''
```
Comment (jwfuller) [inline GitHubAPI.py]: You could set the PAT as an environmental variable or prompt for it and throw an error if it is not set.

File: GitHubAPI.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+import json
+import JIRAloading as JL
+
+issues = JL.JIRALoader('../testJIRA.xml')
+
+# Getting the issues loaded into the JIRALoader object
+issues.get_issue_data()
+
+REPO_OWNER = 'iansincolorado'
+REPO_NAME = 'pythonapis'
```
Comment (jwfuller) [inline GitHubAPI.py]: Could you prompt for these values?

File: README.md [hunk 4]
```diff
@@ -24,6 +24,23 @@ nsidc@nsidc.org for more information.
...
+5. Accept the creation and allow the dependencies to be installed
+6. Activate the new environment with `conda activate JG-ETL` and have fun! 
+
+**Other Conda Notes:**
+- Use `conda --version` to check the version of conda
+- Use `conda env list` to see your current conda environments
+- Use `conda remove -n JG-ETL --all` to remove the environment
+- Use `conda env update -f environment.yml` to update the environment after any changes to the environment.yml file
 
 ## Usage
```
Comment (jwfuller) [inline README.md:48]: Add the steps to run your scripts

File: JIRAloading.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
+import bs4
+import requests
+import re
+
+with open('../testJIRA.xml', 'r') as f:
```
Comment (jwfuller) [inline JIRAloading.py]: Prompt for file name

File: JIRAloading.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    file = f.read()
+
+soup = bs4.BeautifulSoup(file, 'xml')
+
+# Getting all of the items, which are individual JIRA issues
+issues = soup.rss.channel.find_all('item')
+# print(type(issues))
+# print(len(issues))
+# print(issues[0])
```
Comment (jwfuller) [inline JIRAloading.py]: Get rid of debugging statements

File: JIRAloading.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+                checklist = None
+
+            checklists.append(checklist)
+
+        self.checklists = checklists
+        return checklists
+
+
+
+    # This function compiles the data from the titles, links, descriptions, labels, and checklists into a zipped array of tuples, where each tuple is all of the data for one issue.
```
Comment (jwfuller) [inline JIRAloading.py]: Extend comment to include why is it important that they are tuples

File: JIRAloading.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        self.issues_data = issues_data
+        return issues_data
+
+
+
+# Testing
+test = JIRALoader('../testJIRA.xml')
+issues_data = test.get_issue_data()
+# for issue in issues_data:
+    # print(issue[1])
```
Comment (jwfuller) [inline JIRAloading.py]: Is this debugging code?

File: GitHubAPI.py [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+# Only way of authenticating at the current moment is with GitHub PAT tokens, which are deleted from your GitHub account if it is found in a public repository
+PAT = ''
+
+# This function uploads an issue's data to github at the provided repo name, owner, and PAT. (The GitHub PAT user needs to have edit access on the repository.) 
+# It will return the new issue's link and number as a tuple: (link, number). If the issue could not be created, then (None, None) is returned instead
+def upload_issue(issue, REPO_NAME, REPO_OWNER, PAT):
+    URL = f'https://api.github.com/repos/{REPO_OWNER}/{REPO_NAME}/issues'
+    headers = {
+        'Authorization': 'token ' + PAT
+        # 'accept': 'application/vnd.github+json'
```
Comment (jwfuller) [inline GitHubAPI.py]: Do you need this line in the future?

File: GitHubAPI.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        print("HTTP Error")
+        print(errh.args[0])
+    except requests.exceptions.ReadTimeout as errrt:
+        print("Time out")
+    except requests.exceptions.ConnectionError as conerr:
+        print("Connection error")
+    except requests.exceptions.RequestException as errex:
+        print("Exception request")
+    print("Update Success")
+# ****************************************************
```
Comment (jwfuller) [inline GitHubAPI.py]: Could you reuse the code form above?

**Review by jwfuller — CHANGES_REQUESTED** [review]
Awesome start!

**Review by jwfuller — APPROVED** [review]

**Pull request merged**